src/event.rs

```diff
@@ -1,59 +1,411 @@
+            },
+            in_callback: Arc::new(AtomicBool::new(false)),
+            stopped: Arc::new(AtomicBool::new(false)),
         }
     }
 }
+
+impl<S: Clone> Clone for Event<S> {
+    fn clone(&self) -> Self {
+        Event {
+            inner: self.inner.clone(),
+            in_callback: self.in_callback.clone(),
+            stopped: self.stopped.clone(),
+        }
+    }
+}
+
+/// Abstraction over "downgradable" types (i.e., `Rc` and `Arc`).
+pub(crate) trait Downgrade {
+    type Weak;
+
+    fn downgrade(&self) -> Self::Weak;
+}
+
+impl<T> Downgrade for Local<T> {
+    type Weak = LocalWeak<T>;
+
+    fn downgrade(&self) -> Self::Weak {
+        LocalWeak(Rc::downgrade(&self.0))
+    }
+}
+
+impl<T> Downgrade for Sync<T> {
+    type Weak = SyncWeak<T>;
+
+    fn downgrade(&self) -> Self::Weak {
+        SyncWeak(Arc::downgrade(&self.0))
+    }
+}
+
+impl<S: Downgrade> Downgrade for Event<S> {
+    type Weak = Event<S::Weak>;
+
+    fn downgrade(&self) -> Self::Weak {
+        Event {
+            inner: self.inner.downgrade(),
+            in_callback: self.in_callback.clone(),
+            stopped: self.stopped.clone(),
+        }
+    }
+}
+
+impl<T> From<EventInner<T>> for Event<Internal<T>> {
+    fn from(inner: EventInner<T>) -> Self {
+        Event {
+            inner: Internal(inner),
+            in_callback: Arc::new(AtomicBool::new(false)),
+            stopped: Arc::new(AtomicBool::new(false)),
+        }
+    }
+}
+
+impl<T> From<EventInner<T>> for Event<Local<T>> {
+    fn from(inner: EventInner<T>) -> Self {
+        Event {
+            inner: Local(Rc::new(RefCell::new(inner))),
+            in_callback: Arc::new(AtomicBool::new(false)),
+            stopped: Arc::new(AtomicBool::new(false)),
+        }
+    }
+}
+
+impl<T> From<EventInner<T>> for Event<Sync<T>> {
+    fn from(inner: EventInner<T>) -> Self {
+        Event {
+            inner: Sync(Arc::new(Mutex::new(inner))),
+            in_callback: Arc::new(AtomicBool::new(false)),
+            stopped: Arc::new(AtomicBool::new(false)),
+        }
+    }
+}
+
+impl<T> Event<Internal<T>> {
+    fn with_inner<F, O>(&mut self, f: F) -> O
+    where
+        F: Fn(&mut EventInner<T>) -> O,
+    {
+        let ev = &mut self.inner.0;
+        f(ev)
+    }
+
+    pub fn stop(&mut self) -> io::Result<()> {
+        self.set_stopped(true);
+        self.with_inner(|inner| inner.stop())
+    }
+}
+
+impl<T> Event<Local<T>> {
+    fn with_inner<F, O>(&self, f: F) -> O
+    where
+        F: Fn(&mut EventInner<T>) -> O,
+    {
+        let mut ev = self.inner.0.borrow_mut();
+        f(&mut *ev)
+    }
+
+    pub fn stop(&mut self) -> io::Result<()> {
+        self.set_stopped(true);
+        self.with_inner(|inner| inner.stop())
+    }
+}
+
+impl<T> Event<LocalWeak<T>> {
+    fn with_inner<F, O>(&self, f: F) -> O
+    where
+        F: Fn(&mut EventInner<T>) -> O,
+    {
+        let upgraded = self.inner.0.upgrade().unwrap();
```

Comment:
Still reading through but are there guarantees in the `Event<LocalWeak<T>>` and `Event<SyncWeak<T>>` that these unwraps on these `Option<_>`s returned from `upgrade` will be `Some(_)`?

Reply:
Looks like they're only ever being returned for the `spawn_local` and `spawn_sync` variants such that the strong reference is only ever maintained by the `Base`.  So should be good was just trying to understand.

Reply:
I believe it is a guaranteed for the `Local`/single-threaded case since the `Event<Local>` can only be dropped outside of the context of the callback, when it's not running. So in a single call `Drop` will tear everything down before libevent ever has a chance to get to this point.

We spoke offline and I'm definitely not as confident about the `Sync` case; my original idea was to actually make everything an `Arc` and just rely on [`Arc::strong_count`](https://doc.rust-lang.org/beta/std/sync/struct.Arc.html#method.strong_count) under the hood, but ran into some issues with that idea. It's probably worth re-assessing, or just omitting `Sync` altogether until I can guarantee it.